src/main/kotlin/ua/anastasiia/finesapp/service/FineServiceImpl.kt

```diff
@@ -1,55 +1,125 @@
+import ua.anastasiia.finesapp.dto.toMongoViolation
+import ua.anastasiia.finesapp.entity.MongoFine
+import ua.anastasiia.finesapp.exception.CarPlateDuplicateException
 import ua.anastasiia.finesapp.exception.CarPlateNotFoundException
+import ua.anastasiia.finesapp.exception.CarsNotFoundException
 import ua.anastasiia.finesapp.exception.FineIdNotFoundException
-import ua.anastasiia.finesapp.exception.ViolationNotFoundException
-import ua.anastasiia.finesapp.repository.FineRepository
-import ua.anastasiia.finesapp.repository.ViolationRepository
+import ua.anastasiia.finesapp.exception.FinesInLocationNotFound
+import ua.anastasiia.finesapp.exception.NoFinesFoundByDateException
+import ua.anastasiia.finesapp.exception.NoFinesFoundException
+import ua.anastasiia.finesapp.exception.TrafficTicketNotFoundException
+import ua.anastasiia.finesapp.exception.TrafficTicketWithViolationNotFoundException
+import ua.anastasiia.finesapp.repository.MongoFineRepository
+import java.time.LocalDate
 
 @Service
-class FineServiceImpl(
-    val fineRepository: FineRepository,
-    val carService: CarService,
-    val violationRepository: ViolationRepository
-) : FineService {
-    override fun createFine(fineRequest: FineRequest): Fine {
-        val car = carService.getCarByPlate(fineRequest.plate)
-        val fine = Fine(
-            longitude = fineRequest.longitude,
-            latitude = fineRequest.latitude,
-            dateTime = fineRequest.dateTime,
-            photoUrl = fineRequest.photoUrl,
-            car = car.toEntity()
-        )
-        return fineRepository.save(fine)
+@NullableGenerate
+@Suppress("TooManyFunctions")
+class FineServiceImpl(val mongoFineRepository: MongoFineRepository) : FineService {
+    override fun getAllFines(): List<MongoFine> =
+        mongoFineRepository.getAllFines().ifEmpty { throw NoFinesFoundException() }
+
+    override fun getAllFinesInLocation(longitude: Double, latitude: Double, radiusInMeters: Double): List<MongoFine> =
+        mongoFineRepository.getAllFinesInLocation(longitude, latitude, radiusInMeters)
+            .ifEmpty { throw FinesInLocationNotFound(longitude, latitude) }
+
+    override fun getAllFinesByDate(date: LocalDate): List<MongoFine> =
+        mongoFineRepository.getAllFinesByDate(date).ifEmpty { throw NoFinesFoundByDateException(date) }
+
+    override fun getFineById(fineId: ObjectId): MongoFine =
+        mongoFineRepository.getFineById(fineId) ?: throw FineIdNotFoundException(fineId)
+
+    override fun getFineByCarPlate(plate: String): MongoFine =
+        mongoFineRepository.getFineByCarPlate(plate) ?: throw CarPlateNotFoundException(plate)
+
+    @Suppress("SwallowedException")
+    override fun saveFine(fineRequest: FineRequest): MongoFine =
+        runCatching {
+            mongoFineRepository.saveFine(fineRequest.toMongoFine())
+        }.getOrElse {
+            throw CarPlateDuplicateException(fineRequest.car.plate)
+        }
```

Comment:
It's better to check if it is DuplicateKeyException and then throw your custom exception. because there could be multiple reasons to have exception